src/compat/crypto.rs

```diff
@@ -4,73 +4,104 @@
 //
 // Copyright (c) DUSK NETWORK. All rights reserved.
 
+use core::mem::size_of;
+
 use dusk_bls12_381::BlsScalar;
 use dusk_bytes::Serializable;
-use phoenix_core::{Note, PublicKey, ViewKey};
+use phoenix_core::{
+    transaction::{ArchivedTreeLeaf, TreeLeaf},
+    Note, PublicKey,
+};
 
-use alloc::vec::Vec;
+use alloc::{string::ToString, vec::Vec};
 
+use crate::alloc::borrow::ToOwned;
 use crate::{
     key::{self},
     types::{self},
     utils::{self},
     MAX_KEY, MAX_LEN,
 };
 
+const TREE_LEAF_SIZE: usize = size_of::<ArchivedTreeLeaf>();
+
 /// Returns true or false if the note is owned by the index
 /// if its true then nullifier of that note if sent with it
 #[no_mangle]
 pub fn check_note_ownership(args: i32, len: i32) -> i64 {
-    // we just use BalanceArgs again as we don't want to add more cluter types
-    // when the data you want is the same
-    let types::CheckNoteOwnershipArgs { note, seed } =
-        match utils::take_args(args, len) {
-            Some(a) => a,
-            None => return utils::fail(),
-        };
+    let args = utils::take_args_raw(args, len);
+
+    let seed = &args[..64];
+    let leaves: &[u8] = &args[64..];
 
-    let seed = match utils::sanitize_seed(seed) {
+    let seed = match seed.try_into().ok() {
         Some(s) => s,
         None => return utils::fail(),
     };
 
-    let note: Note = match rkyv::from_bytes(&note) {
-        Ok(n) => n,
-        Err(_) => return utils::fail(),
-    };
-
-    let mut is_owned: bool = false;
-    let mut nullifier_found = BlsScalar::default();
-    let mut pk_found: Option<PublicKey> = None;
+    let mut leaf_chunk = leaves.chunks_exact(TREE_LEAF_SIZE);
+    let mut last_pos = 0;
 
-    for idx in 0..=MAX_KEY {
-        let idx = idx as u64;
-        let sk = key::derive_sk(&seed, idx);
-        let vk = ViewKey::from(&sk);
+    let mut notes = Vec::new();
+    let mut nullifiers = Vec::new();
+    let mut block_heights = Vec::new();
+    let mut public_spend_keys = Vec::new();
 
-        if vk.owns(&note) {
-            let nullifier = note.gen_nullifier(&sk);
-
-            nullifier_found = nullifier;
-            is_owned = true;
-            pk_found = Some(PublicKey::from(&sk));
+    for leaf_bytes in leaf_chunk.by_ref() {
+        let TreeLeaf { block_height, note } =
+            match rkyv::from_bytes(leaf_bytes).ok() {
```

Comment:
Is there any reason why you transform a `Result` to an `Option` instead of checking directly the `Result`?

I saw you're doing that in multiple places, not going to comment all of those, just reply here.

Reply:
I will change them in a later PR